### ISSUE TITLE
Features to use only private IP

### ISSUE THREAD

**Issue by jodhi** [kube-hetzner/terraform-hcloud-kube-hetzner#241]
Hello, I just saw the news on [hetzner page regarding flexible networking](https://www.hetzner.com/news/06-22-flexibles-networking/)

And since the kube-hetzner already using private ip by default, i think that should be straightforward to implement worker nodes only with private ip (still not sure).
What do you guys think about worker nodes without public ip?

ps: I can try to create PR if that make sense

**Comment by phaer** [comment]
Good idea, PR more then welcome I think!

**Comment by jodhi** [comment]
cool, i will try this one

**Comment by ricristian** [comment]
And idea would be to extend this from worker nodes also to control plane ones due to the fact that there might be other possibilities to connect to control planes maybe via a VPN and so on . In this way we can basically close completely any chance from receiving requests from outside ...

**Comment by jodhi** [comment]
> An idea would be to extend this from worker nodes also to control plane ones due to the fact that there might be other possibilities to connect to control planes maybe via a VPN and so on . In this way we can basically close completely any chance from receiving requests from outside ...

Yeah, make sense as well, this one will require bastion/jumphost to access the control plane right.
Now you mention it, we need to access the worker node to bootstrap them as well 😓. Looks like we need internal bastion/jumphost to bootstrap the node  🤔

**Comment by ricristian** [comment]
Or at initial bootstrap of the cluster we use public IPv4 and after we finish the job drop those public IP's ?

**Comment by jodhi** [comment]
@ricristian that is smart idea. really

**Comment by ricristian** [comment]
I'm not 100% sure how terraform works but I don't think that we can run in a sequence manner tf modules so a 2nd step can be implemented. I don't see that much hassle for that instead of trying to use jump hosts and so on ...

**Comment by jodhi** [comment]
yeah right, looks like terraform cannot create machine with ip, provision, then destroy it with one go.

what i have in mind 
1. 2-step
```
we put  the machine as usual with public ip in tfvars
terraform apply
then we change the tfvars to remove public ip
terraform apply
```

2. 1-step
```
create control plane nodes with public ip, then create worker nodes without public ip, then get inside the worker nodes through control planes nodes
```

3. 1-step
Use  jumphost, then the terraform module can access all nodes internally. [some docs](https://dev.to/jaysonsantos/connecting-to-services-that-require-jumphost-from-terraform-1aog)

cmiiw. feel free to add

**Comment by ricristian** [comment]
Hmmm this might be something but maybe this flow might be a little bit easier ? 

1st step: 
As you mentioned:
we put  the machine as usual with public ip in tfvars
terraform apply
then we change the tfvars to remove public ip
terraform apply

2. Create both control plane & worker nodes with public IP 
SSH Normal to all those nodes via public IP configure stuff & so on ... 
3. Drop all IPv4 from all VM's ? 


Why I'm trying to avoid jumphosts is because there are a lot of ways to have some kind of jumposts ( VPN, proxy command, Cloudflare, another VM inside Hetzner and so on ) and each one of us might have different use case so maybe it is a global option just to add Public IPv4 when we need to do some stuff on that cluster ( bootrap it, change flavour ) and after that drop those public IP's ( yeah there is a drawback for those few minutes when terraform is doing stuff port 6443 & 22 will be opened in internet if we don't limit it from firewall but still I can afford to have for 2 minutes 2 ports exposed ... it will require some Tesla processing power for someone to break in in under 3 minutes ) ... 

But this are my thoughts hope that it adds some value

**Comment by jodhi** [comment]
thanks @ricristian. nice input, lemme check and try the approach

**Comment by phaer** [comment]
> Or at initial bootstrap of the cluster we use public IPv4 and after we finish the job drop those public IP's ?

I think such approaches run a bit against terraforms spirit. 

In theory, I'd try to implement by using a bastion host for ssh on tcp/22 and kubernetes api on tcp/6443. 
* First step would be a flag for control nodes which provisions the correct sshd config there (probably a bunch of `ProxyJump` directives pointing to all each private ip for nodes in "kubectl get nodes"?).
* At this point you should only need a public ip for one control node of your choice.
* You should already be able to optionally set the flag and add a public ip address for more than one control node. SSH ProxyJump should be stateless server side and the kube api in sync, so you have high availability for "free".
* Second step could be an option to add a tcp load balancer for ssh and kubernetes api which points to private ips of control hosts with the flag set. With that option you'd need to pay for a load-balancer, but get a single public ip which you could firewall off as you like (i.e. only allow your workstation/office ip).

What would you think about such an approach? Is there anything I am missing?
Sadly, I won't find time to implement this myself anytime soon, but I'd be very willing to review a PR and help with working out the details as we go. 
Thanks for raising the topic and your input so far!

**Comment by mysticaltech** [comment]
Thanks, @jodhi, very good idea, and great input @ricristian. For what it's worth, I really like the last iteration on this idea by @phaer. On top of having the nodes only on the private network, it would provide a load balancer for the control plane Kubernetes API, which is something I wanted to implement for a long time.

**Comment by mysticaltech** [comment]
Folks, FYI, slightly related to the above, we now have the possibility to load-balance CP nodes.

**Comment by PurpleBooth** [comment]
I started putting together a spike as to the limitations of public interface-less hosts, and it looks a bit more complicated than simply allowing access via a bastion, we enable the recovery mode to install MicroOS, so we need to change how we install MicroOS in order to allow installation without a public interface, see #261 for my spike.

**Comment by mysticaltech** [comment]
Awesome @PurpleBooth! We can just roll back to using [combustion](https://en.opensuse.org/Portal:MicroOS/Combustion) as we were doing six months ago; this was infinitely more powerful https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/tree/8634a1dc2fd5b05872cf2b35a5aae219ac691282, we just shifted to cloud-init to simplify the code a little.

**Comment by mysticaltech** [comment]
Having given this a second thought, I see two more or less easy ways to implement this:

1. After the agents are deployed and running, access them with a `local-exec`. In it, SSH in with a background process on the server, or through a control-plane node, to permanently disable the public interfaces, meaning edit the networking config file.
2. A little more involved. After the agent nodes run, create snapshots of them, one for each nodepool kind (those tagged with an attributed of private=true, for instance), kill them, and redeploy the snapshots but without a public interface at the Hetzner level.

I would have chosen method 1. What do you think, @jodhi? Would that satisfy your tension?

**Comment by mysticaltech** [comment]
Moving this to a discussion as I do not see it as a priority for the time being.
